compiler/optimizations/gpuTransforms.cpp

```diff
@@ -1302,10 +1302,18 @@ KernelArg::KernelArg(Symbol* symInLoop, GpuKernel* kernel, bool isCompilerGenera
   IntentTag intent = symInLoop->isRef() ? INTENT_REF : INTENT_IN;
   this->formal_ = new ArgSymbol(intent, symInLoop->name, symType);
 
-  if (!isCompilerGeneratedArg &&
-      !isAggregateType(symInLoop->getValType()) &&
+  // scalar values passed to forall/foreach loops by 'ref' intent get "host
+  // registered" so that the gpu can use direct-memory-access on them. Constant
+  // values can safely be passed in by value (so we exclude those). Values
+  // generated by the compiler and temporaries will not be modified by the loop
+  // so we also exclude those.
+  bool isRefIntentScalarArg =
+      !isCompilerGeneratedArg && !isAggregateType(symInLoop->getValType()) &&
       !symInLoop->hasFlag(FLAG_TASK_PRIVATE_VARIABLE) &&
-      !symInLoop->isConstValWillNotChange())
+      !symInLoop->isConstValWillNotChange() && !symInLoop->hasFlag(FLAG_TEMP) &&
+      !symInLoop->hasFlag(FLAG_REDUCTION_TEMP);
```

Comment:
`FLAG_REDUCTION_TEMP` makes sense. We definitely shouldn't be passing it by ref. For `FLAG_TEMP`, I am not really worried, but if you can squeeze a comment saying which test fails if one were to remove it could be helpful.

Reply:
I've added a comment.